docs: add DB migration blog post (#332)
cilium/cilium.io: src/posts/2023-08-28-shekner/index.md

---
path: '/blog/2023/08/28/db-schenker-migration-to-cilium'

> date: '2023-08-28T12:12:00.000Z'

Reviewer: ```suggestion
date: '2023-09-07T12:12:00.000Z'
```

> title: 'How the IT for Land Business of DB Schenker Migrated from Calico to Cilium'

Reviewer: ```suggestion
title: 'How DB Schenker Migrated from Calico to Cilium'
```

isPopular: false
isFeatured: false
ogImage: only-cilium.png
ogSummary: 'Learn how DB Schneker migrated their Kubernetes platform CNI from Calico to Cilium'
categories:
  - Community
  - Technology
tags:
  - Cilium
---

_August 28th, 2023_

> _Author: Amir Kheirkhahan, BD Schneker_

Reviewer: ```suggestion
_Author: Amir Kheirkhahan, DB Schneker_
```

![only Cilium](only-cilium.png)

## History behind the Migration to Cilium

> In the past, the land business of [DB Schenker](https://www.dbschenker.com/de-de) used Calico as a Container Network Interface ([CNI](https://www.cni.dev/)) for in-Kubernetes-Cluster communication, like pod-to-pod communication. Recently our team had the chance to participate in the 2023 [KubeCon](https://events.linuxfoundation.org/kubecon-cloudnativecon-europe/) in Amsterdam, where we learned a lot about [eBPF](https://ebpf.io/) and especially [Cilium](https://cilium.io/), which was an important driver for us to question our CNI strategy.

Reviewer: ```suggestion
In the past, the IT unit for the land transportation business of [DB Schenker](https://www.dbschenker.com/de-de) used Calico as a Container Network Interface ([CNI](https://www.cni.dev/)) for in-Kubernetes-Cluster communication, like pod-to-pod communication. Recently our team had the chance to participate in the 2023 [KubeCon](https://events.linuxfoundation.org/kubecon-cloudnativecon-europe/) in Amsterdam, where we learned a lot about [eBPF](https://ebpf.io/) and especially [Cilium](https://cilium.io/), which was an important driver for us to question our CNI strategy.
```

Another important consideration was the kube-proxy replacement feature. When a new service is created, the API server notifies all kube-proxy agents running on the worker nodes about this event. Then kube-proxy creates iptable rules to make sure each packet destined for a service is forwarded to the one of the correct backend pods. It is the same for a new pod when the endpoints in iptables are updated.  In the following migration steps, we kept Cilium’s kube-proxy replacement option disabled to reduce the complexity of the migration.

Finally, we also needed to think about our Kafka clusters. We run Kafka inside our Kubernetes environment by using the [strimzi](https://strimzi.io/) chart which includes very sensitive and important data. To ensure the stability of Kafka, we needed  to prevent losing the cluster state or running into a Kafka leader election issue.

> To get a better understanding of our infrastructure, I’ll shortly point out some tooling and approaches which we use within DB Schenker. We use self-managed Kubernetes clusters on AWS where the worker and controller nodes are running independently on separate VMs. The worker nodes run on a mix of spot and on-demand instances. The base Images are built by Packer and Ansible and the whole infrastructure setup is managed by Terraform. To meet IT Security requirements, every day we throw away a set of our worker and controller nodes. We also have automated processes to rotate all our nodes. We do this by starting a new node and waiting until it is ready, then we move all workloads running on the old node away from it before terminating the old node in AWS as well as in Kubernetes.  These steps are repeated for each of our worker pools.

Reviewer: ```suggestion
To get a better understanding of our infrastructure, I’ll shortly point out some tooling and approaches which we use within DB Schenker. We use self-managed Kubernetes clusters on AWS where the worker and controller nodes are running independently on separate VMs. The worker nodes run on a mix of spot and on-demand instances. The base Images are built by Packer and Ansible and the whole infrastructure setup is managed by Terraform. To meet IT Security requirements, every day we throw away a set of our worker and controller nodes. We also have automated processes to rotate all our nodes. We do this by starting a new node and waiting until it is ready, then we drain the old node and move all workloads away from it before terminating the old node in AWS as well as in Kubernetes. These steps are repeated for each of our worker pools.
```

Now it is the time to install Cilium where the configuration looks like this:

```yaml
# migration, for connectivity between calico and cilium

  bpf:

    hostLegacyRouting: true 

# Set cluster name. It would be interesting for cluster mesh

  cluster:

    id: 0

    name: prod-cluster

  cni:

# migration, disable CNI changes

    customConf: true 

# should be always false so it doesn't remove cilium

    uninstall: false 

# Assign a new CIDR for cilium

 ipam:

    operator:

      clusterPoolIPv4PodCIDRList:

        - 10.x.x.x/16          

# migration, otherwise all calico pods are restarted and we get downtime

  operator:

    unmanagedPodWatcher: 

      restart: false

# migration

  policyEnforcementMode: never 

# We stay with kubeproxy

  kubeProxyReplacement: disabled  

# change the tunnel port, If you have already calico with VXLAN

  tunnel: vxlan

>   tunnelPort: 8472

Reviewer: ```suggestion
  tunnelPort: 8473
```

```

If you check the Cilium status, you will see all nodes are managed by the Cilium CNI. Now we can remove all Calico CRDs and other leftovers.

```bash
    kubectl delete crd bgpconfigurations.crd.projectcalico.org bgppeers.crd.projectcalico.org blockaffinities.crd.projectcalico.org caliconodestatuses.crd.projectcalico.org clusterinformations.crd.projectcalico.org felixconfigurations.crd.projectcalico.org globalnetworkpolicies.crd.projectcalico.org globalnetworksets.crd.projectcalico.org hostendpoints.crd.projectcalico.org ipamblocks.crd.projectcalico.org ipamconfigs.crd.projectcalico.org ipamhandles.crd.projectcalico.org ippools.crd.projectcalico.org ipreservations.crd.projectcalico.org kubecontrollersconfigurations.crd.projectcalico.org networkpolicies.crd.projectcalico.org networksets.crd.projectcalico.org
    kubectl delete -n kube-system deploy calico-kube-controllers
    kubectl delete -n kube-system ds/calico-node
    kubectl delete -n kube-system cm/calico-config
    kubectl delete -n kube-system ciliumnodeconfig cilium-default
```

### Step 8- Post migration steps 

> As the [official documentation](https://docs.cilium.io/en/stable/installation/k8s-install-migration/#post-migration) suggested, we do some post migration steps, like restarting the unmanaged pods by cilium operator, enforcing NetworkPolicy, and disabling eBPF host routing which could cause short interruption.

Reviewer: ```suggestion
As the [official documentation](https://docs.cilium.io/en/stable/installation/k8s-install-migration/#post-migration) suggested, we do some post migration steps, like restarting the unmanaged pods by cilium operator, enforcing NetworkPolicy, and enabling eBPF host routing which could cause short interruption.
```

---
path: '/blog/2023/09/07/db-schenker-migration-to-cilium'
date: '2023-09-07T12:12:00.000Z'
title: 'How DB Schenker Migrated from Calico to Cilium'

> isPopular: false

Reviewer: ```suggestion
isPopular: true
```

isFeatured: false
ogImage: after-removing-calico-resources.png
ogSummary: 'Learn how DB Schneker migrated their Kubernetes platform CNI from Calico to Cilium'
categories:
  - Community
  - Technology
tags:
  - Cilium
---

> _Author: Amir Kheirkhahan, DB Schneker_

Reviewer: ```suggestion
_Author: Amir Kheirkhahan, DB Schenker_
```

Another important consideration was the kube-proxy replacement feature. When a new service is created, the API server notifies all kube-proxy agents running on the worker nodes about this event. Then kube-proxy creates iptable rules to make sure each packet destined for a service is forwarded to the one of the correct backend pods. It is the same for a new pod when the endpoints in iptables are updated.  In the following migration steps, we kept Cilium’s kube-proxy replacement option disabled to reduce the complexity of the migration.

Finally, we also needed to think about our Kafka clusters. We run Kafka inside our Kubernetes environment by using the [strimzi](https://strimzi.io/) chart which includes very sensitive and important data. To ensure the stability of Kafka, we needed  to prevent losing the cluster state or running into a Kafka leader election issue.

> To get a better understanding of our infrastructure, I’ll shortly point out some tooling and approaches which we use within DB Schenker. We use self-managed Kubernetes clusters on AWS where the worker and controller nodes are running independently on separate VMs. The worker nodes run on a mix of spot and on-demand instances. The base Images are built by Packer and Ansible and the whole infrastructure setup is managed by Terraform. To meet IT Security requirements, every day we throw away a set of our worker and controller nodes. We also have automated processes to rotate all our nodes. We do this by starting a new node and waiting until it is ready, then we drain the old node and move all workloads away from it before terminating the old node in AWS as well as in Kubernetes. These steps are repeated for each of our worker pools.

Reviewer: ```suggestion
To get a better understanding of our infrastructure, I’ll shortly point out some tooling and approaches which we use within DB Schenker. We use self-managed Kubernetes clusters on AWS where the worker and controller nodes are running independently on separate VMs. The worker nodes run on a mix of spot and on-demand instances. The base Images are built by Packer and Ansible and the whole infrastructure setup is managed by Terraform. To meet IT Security requirements, every day we throw away a set of our worker and controller nodes. We also have automated processes to rotate all our nodes. We do this by starting a new node and waiting until it is ready, then we drain the old node and move all workloads away from it before terminating the old node in our cloud provider as well as in Kubernetes. These steps are repeated for each of our worker pools.
```

![After Starting the Rotation](after-starting-the-rotation.png)

<div align="center" style="font-style:italic">Figure 3. After Starting the rotation, Node C starts with both CNIs and Cilium has CNI ownership. Node C currently has Cilium configurations in place which is discussed later at step 1 of migration</div>

![Draining the oldest worker node](draining-the-oldest-worker-node.png)

<div align="center" style="font-style:italic">Figure 4. Draining the older worker node (here Node A) and re-schdule the pods on the new Node</div>

![Node A fully replaced](node-a-fully-replaced.png)

<div align="center" style="font-style:italic">Figure 5. Node A is removed from the LoadBalancer so that it no longer receives traffic, and then terminated. Node A is fully replaced by Node C</div>

![After the Full rotation](after-the-full-rotation.png)

<div align="center" style="font-style:italic">Figure 6. After the full rotation of both Nodes, both CNIs should be running on new nodes with Cilium leading</div>

![After removing calico resources](after-removing-calico-resources.png)

> <div align="center" style="font-style:italic">Figure. 7. After removing Calico resources and clean up the lables, the new nodes have only Cilium installed on it</div>

Reviewer: ```suggestion
<div align="center" style="font-style:italic">Figure. 7. After removing Calico resources and clean up the labels, the new nodes have only Cilium installed on it</div>
```

![only-cilium](only-cilium.png)

<div align="center" style="font-style:italic">Figure 8. At the end, only Cilium is running on all nodes.</div>

## Live migration

### Step 1- Preparation

Before starting the migration, we need to make sure that we have proper back-ups of the workloads. In the Kubernetes context, the most important piece is to keep etcd backedup in a safe place, like an AWS S3 Bucket. In case the migration fails, we have a backup in place to perform a disaster recovery to the state before the migration.

> The next thing that should be prepared is building the migration image (AMI in AWS context). Some of the important changes are:

Reviewer: ```suggestion
The next thing that should be prepared is building the migration base image. Some of the important changes are:
```

- Add the label `"io.cilium.migration/cilium-default=true"` to all newly created nodes.  As soon as a new node is launched and has this label, the predefined Cilium config will be applied to this node. These predefined configs are declared in the [CiliumNodeConfig](https://docs.cilium.io/en/latest/configuration/per-node-config/#ciliumnodeconfig-objects) object.

- Add the Taint `"node.cilium.io/agent-not-ready=:NoSchedule.` to all newly created nodes. This prevents a race condition between the CNIs that are now running in parallel on the new nodes. Since Calico is sometimes faster in starting, the node gets marked as ready and newly scheduled pods on the node get IP addresses assigned by Calico. With the taint, Cilium is able to start up properly and then take over assignment of IPs.
  For more information refer to [official Cilium documentation.](https://docs.cilium.io/en/latest/installation/taints)

> You should make your changes ready for deployment after Cilium installation. In our case, we built a new AMI with these changes and kept them ready to deploy. We used Terraform for deploying the infrastructure in AWS and in this context, we rolled out the new images right after the Cilium installation. With that all new nodes have this label and taint and Cilium is the leading CNI on those nodes.

Reviewer: ```suggestion
You should make your changes ready for deployment after Cilium installation. In our case, we built a new AMI with these changes and kept them ready to deploy. We used Terraform for deploying the infrastructure in our cloud environment and in this context, we rolled out the new images right after the Cilium installation. With that all new nodes have this label and taint and Cilium is the leading CNI on those nodes.
```

### Step 3- Cilium installation

Afterwards, we apply the changes to Terraform that contain the labels and taints that are required to tell Cilium to run as leading CNI on those nodes. It is important to mention that this step contains no new node deployment and is rather a preparation for the final installation and configuration of Cilium. Node deployment happens during the rotation.

Now it is the time to install Cilium where the configuration looks like this:

```yaml
# migration, for connectivity between calico and cilium

  bpf:

    hostLegacyRouting: true 

# Set cluster name. It would be interesting for cluster mesh

  cluster:

    id: 0

    name: prod-cluster

  cni:

# migration, disable CNI changes

    customConf: true 

# should be always false so it doesn't remove cilium

    uninstall: false 

# Assign a new CIDR for cilium

 ipam:

    operator:

      clusterPoolIPv4PodCIDRList:

        - 10.x.x.x/16          

# migration, otherwise all calico pods are restarted and we get downtime

  operator:

    unmanagedPodWatcher: 

      restart: false

# migration

  policyEnforcementMode: never 

# We stay with kubeproxy

  kubeProxyReplacement: disabled  

> # change the tunnel port, If you have already calico with VXLAN

Reviewer: ```suggestion
# change the tunnel port, If you have already calico with VXLAN
```
```suggestion
# Use Cilium suggested tunnel port 8473
```

```bash
    kubectl annotate pod -n kube-system -l k8s-app=kube-dns cluster-autoscaler.kubernetes.io/safe-to-evict=false
```

Now if you check the status of Cilium, you will still see that no pods are managed by Cilium, except `_coredns_` pods. This is expected. Calico is still running on the node and has CNI ownership. The nodes should be drained, cordoned, and the new nodes created as described  later.

### Step 5- Scale up Kafka

> Now you can scale up Kafka’s underlying worker pools to the previous numbers in AWS. New nodes should have both CNIs running, but Cilium will take over leadership. After that you can start Kafka applications again. First start Zookeeper and wait for it to be ready and then start the Kafka brokers.

Reviewer: ```suggestion
Now you can scale up Kafka’s underlying worker pools to the previous numbers in your cloud provider. New nodes should have both CNIs running, but Cilium will take over leadership. After that you can start Kafka applications again. First start Zookeeper and wait for it to be ready and then start the Kafka brokers.
```

    kubectl delete -n kube-system deploy calico-kube-controllers
    kubectl delete -n kube-system ds/calico-node
    kubectl delete -n kube-system cm/calico-config
    kubectl delete -n kube-system ciliumnodeconfig cilium-default
```

### Step 8- Post migration steps 

As the [official documentation](https://docs.cilium.io/en/stable/installation/k8s-install-migration/#post-migration) suggested, we do some post migration steps, like restarting the unmanaged pods by cilium operator, enforcing NetworkPolicy, and enabling eBPF host routing which could cause short interruption.

```yaml
   bpf:
     hostLegacyRouting: false
   cni:
     customConf: false
   operator:
     unmanagedPodWatcher:
       restart: true
   policyEnforcementMode: default
```

We now have all new nodes only with Cilium as the CNI. We removed Calico resources and Cilium migration labels from the code. As mentioned earlier, we have a daily automatic rotation approach for each worker pool and hence all nodes are replaced gradually and Cilium will be the only CNI available on those nodes.

## Conclusion

> This blog post walked through how we migrated from Cilium to Calico using the new `CiliumNodeConfig` feature. Overall, we found the process to be smooth and are excited for the possibility that switching to Cilium will unlock for our platform.

Reviewer: ```suggestion
This blog post walked through how we migrated from Calico to Cilium using the new `CiliumNodeConfig` feature. Overall, we found the process to be smooth and are excited for the possibility that switching to Cilium will unlock for our platform.
```
